Clarify ros2 arguments precedence (#289)
ros2/design: articles/150_ros_command_line_arguments.md

As is, this enclave assignment applies to each and every Domain Participant that `some_ros_executable` spawns unless explicitly ignored in code or overridden via security environment variables.  

#### Precedence order of command line arguments

> The following precedence order applies to parameter assignments:

Reviewer: @ivanpauno nit: consider using H5 headers to separate remapping rules precedence from parameter assignment precedence.

* Arguments targeting a specific node prevail arguments targeting all the nodes in the executable.

> * Arguments are parsed from the leftmost to the rightmost. In case of contradictory assignments, the rightmost argument prevails.

Reviewer: @ivanpauno 
```suggestion
* Arguments are parsed from the leftmost to the rightmost. In case of overlapping assignments, the rightmost argument prevails.
```

> * Arguments are parsed from the leftmost to the rightmost. In case of contradictory assignments, the rightmost argument prevails.

Reviewer: @ivanpauno also
```suggestion
* Arguments are parsed left to right. In case of contradictory assignments, the rightmost argument prevails.
```

Reviewer: I think it'd be also be nice to include a rationale for each precedence.

Author: This is stating how it currently works in order to kick off a discussion on the topic.
This is not trying to rationalize why it works in the way it works.

As an example:

```sh
ros2 run some_package some_exec --ros-args -r my_topic:=asd -r my_topic:=bsd -p my_node:my_param:=1 -p my_param:=2 -p another_param:=1 -p another_param:=2
```

> Supposing the executable only has a node named `my_node`, this will result in `my_param` being set to `1` and `another_param` being set to `2`.

Reviewer: @ivanpauno 
```suggestion
Assuming that `some_exec` only has a node named `my_node`, this will result in `my_param` being set to `1` and `another_param` being set to `2`.
```

The following precedence order applies to remapping rules:

> * Arguments are parsed from the leftmost to the rightmost. In case of contradictory assignments, the leftmost argument prevails.

Reviewer: @ivanpauno 
```suggestion
* Arguments are parsed from left to right. In case of overlapping rules, the leftmost argument prevails.
```

* Arguments targeting a specific node DO NOT prevail arguments targeting all the nodes in the executable.

As an example:

```sh
ros2 run some_package some_exec --ros-args -r my_topic:=first -r my_topic:=second -r my_node:another_topic:=first -r another_topic:=second
```

> Supposing the executable only has a node named `my_node`, this will result in `my_topic` being remapped to `first` and `another_topic` being remapped to `first`.

Reviewer: @ivanpauno 
```suggestion
Assuming that `some_exec` only has a node named `my_node`, this will result in `my_topic` being remapped to `first` and `another_topic` being remapped to `first`.
```

In the case of parameter files, the following precedence apply:

> * Assignments targeting a specific node prevail to wildcard assignments.

Reviewer: @ivanpauno 
```suggestion
* Parameter targeting a specific node prevail over those using wildcards.
```

Author: I will apply the following diff, which I find clearer:
```suggestion
* Parameter assignments targeting a specific node prevail over those using wildcards.
```

> * Within a file, in case of the contradictory assignments, the last assignment applies (i.e.: the one nearer to the end of the file).

Reviewer: @ivanpauno 
```suggestion
* Within a file, in case of the overlapping assignments, the last assignment applies (i.e.: the one closest to EOF).
```

Remove the need for double dash tokens (`--`), conventionally used to signify the end of CLI options for a command, by adding the `--ros-` prefix to all ROS specific command line flags e.g. `--ros-remap`, `--ros-param`, etc.
In exchange, it makes argument extraction slightly more difficult as all options must be known ahead of time, whereas `--ros-args`-based namespacing can achieve the same with a couple rules.
It also increases command line verbosity.

## Precedence order discussion

> There are some issues of the current precedence rules:

Reviewer: @ivanpauno 
```suggestion
There are some issues with the current precedence rules:
```
Also, when you say *current* we need to be specific about when.

Author: > Also, when you say current we need to be specific about when.

With current, I mean the ones described above.
After a discussion on how the precedence rules should be, I will update this again with the conclusion.

* Precedence orders for remapping rules and parameter assignments are inconsistent.
* In the past, inconsistencies were detected between `rclcpp` and `rclpy` ([example issue](https://github.com/ros2/rclcpp/issues/953)).
* Weird interaction with launch files.

Example of the last item:

```python
LaunchDescription([
  Node(
    package='my_pkg',
    exec='my_exec',
    name='my_node',
    parameters=['/path/to/my_params.yaml', {'my_int': '3'}]
  )
])
```

> Where `my_params.yaml` can be find in *Precedence order of command line arguments* section.

Reviewer: @ivanpauno I'd reproduce the example parameter file here.

Author: I will address this after we reach a conclusion on how precedence should work.
As my current goal is to start a discussion, having a clear reference to the file isn't important.

> The parameter file is assigning `my_int` to `2`.

Reviewer: @ivanpauno 
```suggestion
The parameter file is assigning `2` to `my_int`.
```

Launch cannot know the fully qualified name of the node at launch time, as the node namespace wasn't specified.
Thus, it needs to use a "wildcard" rule for the assignment made in the dictionary, and it won't override the parameter file assignment.

Before https://github.com/ros2/launch_ros/pull/154 was merged, it was impossible to override a parameter assignment in a parameter file that targeted an specific node from a launch file.
Now it's possible by doing:

```python
LaunchDescription([
  Node(
    package='my_pkg',
    exec='my_exec',
    name='my_node',
    namespace='my_ns',
    parameters=['/path/to/my_params.yaml', {'my_int': '3'}]
  )
])
```

As the fully qualified node name is known at launch time in this last case, `my_int` will be set to `3`.

Possible solutions:
* Warn when parameter files and parameter dictionaries are combined in `Node`, and the fully qualified node name is not known.

> * Modify precedence between wildcard and not wildcard assignments.

Reviewer: @ivanpauno shouldn't this design doc weigh pros and cons and make an executive decision?

Author: as commented previously, I'm not trying to weight advantages and disadvantages at the moment.
